create first part of backend

chyVacheck opened the pull request:
# Backend "level-1"
- Серверная часть написана и выложена по адресу в readme на сервер
- Запросы get post delete patch работают
- SSL ключ установил 18.01.23
___
- cors выключен, но заготовка под него есть, а именно в файлах создан объект куда останется вписать разрешенные адреса

chyVacheck committed:
create backend

chyVacheck committed:
refactor: убрать eslint-disable в коде

Убрать eslint-disable, и исправить ошибки которые возникают

chyVacheck committed:
feat: переделать из mesto в movies

chyVacheck committed:
Update README.md

chyVacheck committed:
Update index.js

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [ ] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [ ] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
  **Присутствуют блокировки вида `eslint-disable`.**
- [ ] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
**Ошибок не должно быть без блокировок.**
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [ ] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
**Не защищен роут `/signout`.**
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [ ] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [ ] сообщение об ошибке соответсвует ее типу;
	**Не обработана ошибка дубля email при редактировании пользователя.**
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
  **1. Поле `year` при создании фильма должно валидироваться как строка.**
  **2. Поле `movieId` при создании фильма должно валидироваться как число + должно быть обязательным.**
  **3. Параметр movieId при удалении фильма должен валидироваться как `_id` базы `mongo`.**
  **4. Поля при редактировании пользователя должны быть обязательными + должны валидироваться.**
  - [ ] Данные валидируются перед добавлением в базу.
  **1. Поле `year` на уровне схемы должно валидироваться как строка.**
  **2. Поля `image`, `trailerLink`, `thumbnail` должны валидироваться на уровне схемы.**
  **3. Поле `movieId` на уровне схемы должно валидироваться как число.**
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [ ] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [ ] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.
  **Отсутсвтует адрес базы для режима разработки.**

### Хорошие практики
- [ ] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
**Роуты подключены отдельно.**
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--2.14--> Валидация описана в отдельном модуле.
**Валидация описана в роутах.**
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [ ] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone
**`api` присутствует в названии, но не является поддоменом.**

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [ ] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  **Адрес для режима разработки отсутствует.**
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Присутствуют сообщения, прописанные вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [ ] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
**`api` присутствует в названии, но не является поддоменом.**
- [ ] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.
**Адрес для режима разработки отсутствует.**

## Количество баллов: 59

## Доп. Комментарии:
1. **У фильма нет поля `createdAt`.**
![2023-01-19_17-40-15](https://user-images.githubusercontent.com/86304001/213433376-ded17c71-e7d7-4e29-9735-9e4ca092e905.png)
2. **При создании фильма можно обойтись без паровоза деструктуризации:**
```js
Movie.create({ ...req.body, owner: req.user._id })
```
3. **Лишний `catch`. Метод send не возвращает промис.**
![2023-01-19_17-42-54](https://user-images.githubusercontent.com/86304001/213433918-772645eb-51bb-4298-b975-0cc7bb563384.png)
4. **Не нужно делать кастомную проверку на русс  и англ имена. Этого нет в ТЗ.**

chyVacheck committed:
fix: добавить адрес для режима разработки

chyVacheck committed:
refactor fix: защитить signout, перенести все роутеры в 1

1. Теперь без авторизации нельзя перейти по адресу signout
2. Теперь все роутеры находятся в одном файле

chyVacheck committed:
refactor: убрать не нужные вещи, убрать деструкторизацию

chyVacheck committed:
refactor: убрать eslint-disable

chyVacheck committed:
Update constants.js

chyVacheck committed:
feat: добавить проверку на дублирование почты в роутер

1. Теперь при попытке поменять почту на почту другого пользователя, будет возвращаться ошибка дупликации данных

chyVacheck committed:
fix: разобраться с данными + сделать проверку ссылок

1. Сделать проверку данных на URL
2. Поменять типы данных

chyVacheck committed:
refactor: вынести в constants информационные сообщения

1. Теперь в constants есть сообщение INCORRECT_DATA

chyVacheck commented:
- Переместить все роутеры в 1 файл + защитить signout
- Убрать eslint-disable, + убрать деструкторизацию, там где не нужно
- Добавить проверку почты при запросе пользователя patch (что бы он не поменял на почту другого пользователя)
- Добавить проверку данных на URL (при запросах на создание фильма)
- Немного обновил constants
- Добавил адрес mongodb для режима разработки
- Мелкие исправления по типу заменить сообщения на переменные из констант, поменять типы данных в моделях и т.д.

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [ ] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [ ] сообщение об ошибке соответсвует ее типу;
	**Ошибки выглядят следующим образом:**        
![2023-01-20_16-17-00](https://user-images.githubusercontent.com/86304001/213671502-35d1ba06-8054-4c6f-b530-55d05e4ecc9a.png)
  - [x] асинхронные обработчики завершены блоком catch;
  - [ ] API не возвращает стандартных ошибок базы данных или Node.js.
  **Так как ошибки не обрабатываются, может прилететь стандартная ошибка.**
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
  **1. Параметр `movieId` при удалении фильма должен валидироваться как `_id` базы `mongo`.**
  **2. Поля при редактировании пользователя должны быть обязательными + должны валидироваться в соответствии с ТЗ.**
  - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--2.14--> Валидация описана в отдельном модуле.
**Валидация описана в роутах.**
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [ ] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
**Ошибки не обрабатываются.**
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [ ] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone
**`api` присутствует в названии, но не является поддоменом.**

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [ ] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  **Адрес прописан в `app.js`.**
  - [x] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [ ] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
**`api` присутствует в названии, но не является поддоменом.**
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 78

chyVacheck committed:
feat: разрешить использовать console

chyVacheck committed:
Create default.json

chyVacheck committed:
fix: исправить проблему неправильного ответа ошибок

chyVacheck committed:
refactor: подключить config.json

1. Подключить config.json, в котором находятся данные для запуска
2. Убрать лишние комментарии
3. Сделать логику включения cors через  сonfig.json
4. Переименовать переменную для адреса БД

chyVacheck committed:
refactor: заменить на константы

chyVacheck committed:
Update constants.js

chyVacheck committed:
refactor: переместить валидацию в 1 файл

1. Валидация собрана в один файл
2. Валидация в роутерах заменена на методы из главного файла валидации

chyVacheck committed:
feat: поменять путь до сервера

1. В config.json добавлен адрес на который будут идти все запросы
2. Все пути, теперь начинаются с "/api", любой другой как и раньше возвращает ошибку 404

chyVacheck commented:
## Исправления 
- Переместить адрес к БД в config.json
- Исправить возврат ошибок, теперь это снова json объект, а не html страница
## Нововведения
- Поменять путь, теперь все запросы начинаются на url/api, где url - доменное имя
- Валидация была перенесена в отдельный файл
- Порт теперь прописан не в app.js, а в config.json (речь идет о порте для разработки, для продакшина все также в .env)
- Изменены условия для входящих параметров
### Мелкие изменения
- Часть переменных была заменена на константы
- Для имени добавлены ограничения по длине
#### Более точно смотрите в комитах

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
  - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
  - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [x] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 100

## Доп. Комментарии:
1. **Не стоит вызывать следующий мидлвар посреди текущего.**
![2023-01-21_18-52-57](https://user-images.githubusercontent.com/86304001/213867677-cbe8a9a6-58f2-4e51-965c-a07d40743f47.png)
2. В `login` секрентный ключ прописан вручную.

chyVacheck committed:
fix: переместить вызов next в конец

chyVacheck committed:
refactor: убрать "хардкорный" ключ из функции авторизации

chyVacheck commented:
## Исправления
- Отсутствуют
## Нововведения
- Отсутствуют
## Мелкие изменения
- Убрать "хардкорный" ключ из функции авторизации (теперь ключ в режиме разработки берется из config.json)
- Убрать вызов next из середины, и переместить его в конец
#### Более точно смотрите в комитах
